### PR TITLE
feat: create schemas & credentials for oft ctes & kdes

### PR DESCRIPTION
**Pull request by rhofvendahl**
This PR creates several schemas & corresponding certificates representing the Critical Tracking Events (CTEs) & Key Data Elements (KDEs) which ag supply chain organizations will be required to report once the upcoming FSMA Traceability Rule takes effect.

Outline of the FSMA Proposed Rule for Food Traceability: https://www.fda.gov/food/food-safety-modernization-act-fsma/fsma-proposed-rule-food-traceability

Details about KDEs required for reporting on the various CTEs: https://www.fda.gov/food/food-safety-modernization-act-fsma/which-key-data-elements-would-apply-me

Mock data for reporting on an example product: https://www.fda.gov/media/145838/download

### CODE REVIEW

**Review by mprorock — APPROVED** [review]
Awesome work

**Comment by OR13** [comment]
woah.

File: docs/openapi/components/schemas/common/FSMAAbstractKDE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAAbstractKDE.yml:27]:
```suggestion
      '@id': https://schema.org/name
```

File: docs/openapi/components/schemas/common/FSMACreatingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMACreatingCTE.yml:27]: prefer `location`.... you are repeating `Created` a lot....

File: docs/openapi/components/schemas/common/FSMACreatingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMACreatingCTE.yml:40]: id type confusion.

Comment (OR13) [reply docs/openapi/components/schemas/common/FSMACreatingCTE.yml:40]: Find a `property` from this list: 

https://schema.org/Date

for example:

https://schema.org/dateCreated
https://schema.org/productionDate

File: docs/openapi/components/schemas/common/FSMACreatingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMACreatingCTE.yml:66]: sadly... not a unit code... https://schema.org/unitCode

Comment (OR13) [reply docs/openapi/components/schemas/common/FSMACreatingCTE.yml:66]: I think we have this same problem on several schemas.

File: docs/openapi/components/schemas/common/FSMAFirstReceiverData.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAFirstReceiverData.yml:41]: id / type confusion

File: docs/openapi/components/schemas/common/FSMAFirstReceiverData.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAFirstReceiverData.yml:62]: id / type confusion

File: docs/openapi/components/schemas/common/FSMAGrowingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAGrowingCTE.yml:28]: move TODO to an issue...

File: docs/openapi/components/schemas/common/FSMAGrowingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAGrowingCTE.yml:33]: id / type confusion

File: docs/openapi/components/schemas/common/FSMAGrowingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAGrowingCTE.yml:72]: is this a real email?

File: docs/openapi/components/schemas/common/FSMAGrowingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAGrowingCTE.yml:70]: You are gonna end up with a bunch of undefined terms... thats going to be data that is not processable or mergeable...

File: docs/openapi/components/schemas/common/FSMAGrowingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAGrowingCTE.yml:73]: lol `-` vs `.`

File: docs/openapi/components/schemas/common/FSMAProduct.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAProduct.yml:34]: id / type confusion.

File: docs/openapi/components/schemas/common/FSMAProduct.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAProduct.yml:114]: why you include an `N/A` here... ?

File: docs/openapi/components/schemas/common/FSMAProduct.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAProduct.yml:85]: This is maybe not how I would model this... I would be happy to learn more about this use case though.

File: docs/openapi/components/schemas/common/FSMAReceivingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAReceivingCTE.yml:33]: id / type confusion

File: docs/openapi/components/schemas/common/FSMAReceivingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAReceivingCTE.yml:42]: id / type confusion

File: docs/openapi/components/schemas/common/FSMAReceivingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAReceivingCTE.yml:74]: You will have lots of undefined terms

File: docs/openapi/components/schemas/common/FSMAReceivingCTE.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAReceivingCTE.yml:131]: interesting semantics, why not `source` and `destination`?

File: docs/openapi/components/schemas/common/FSMAShipment.yml
Comment (OR13) [inline docs/openapi/components/schemas/common/FSMAShipment.yml:41]: id / type... what you mean to use here is:

The property with `id`: https://schema.org/location
Which supports the `type: https://schema.org/Place

This mistake is all over the place (not just in your PR here, but in many other schemas).

Here is an example of what I mean:

https://v.jsld.org/6DxQAm4Rf3J61R2L8HSynh9Hg741hKe4F4vJU1Z7VbdYnDPyx3WyzjuBcr5UMHpLtbis7Xsx7AjbWgYqaPK4fNh86qdiqvaJ7eN7GTTvRPiYoRR3aYziZ6z89KrYXDMCE1116SiKGXynhVJzvWBHcotgSemPyzaETLWCX5z2UeeMxsuRmgnn7s1qf8xRMaXDqT41EoDEZcps3L61c33oxjtqZZduTDhWD2Sw6xDMEYZATLmrh1rCE65Xwj6Ahpg8cpFmQXUVPygtAPShnZwbKXuzCZPfo9tYGkksEiNbrNLKB6YXaaKEdiDvxSXZaqodyF9nacxkgR58VB2Yn8PcRK2xvHZ2KH8JhmvPrjT8K9yYJ8riev2s9npwsGie2HgY2kxgpuKz6nFS2Ddad2QTxVBKGKR2eZjZXQGpc4wsY7qojJkCukn85Zj49Pw5kyPsLJiiqWnKpgRiW8

Comment (OR13) [reply docs/openapi/components/schemas/common/FSMAShipment.yml:41]: The example is taken from: https://schema.org/Place#eg-0171

Comment (OR13) [reply docs/openapi/components/schemas/common/FSMAShipment.yml:41]:
```
_:c14n0 <https://schema.org/location> _:c14n3 ._:c14n0 <https://schema.org/location> _:c14n3 .
...
_:c14n3 <https://schema.org/address> _:c14n2 .
... 
_:c14n2 <https://schema.org/postalCode> "80209" .
```

Comment (OR13) [reply docs/openapi/components/schemas/common/FSMAShipment.yml:41]: You only get "machine readability" if your JSON-LD /  RDF is correct... that means using `id` and `type` correctly.

**Review by OR13 — APPROVED** [review]
This PR is massive.

I am approving it in the interest of seeing SMALLER PRs which refine the various errors which I have commented on, which will remain observable after it is merged.

The main thing to learn about is here:

https://github.com/w3c-ccg/traceability-vocab/pull/479/files#r922547127

After this PR is merged, this chart will show a massive amount of failure:

https://w3c-ccg.github.io/traceability-vocab/#certificates-with-undefined-terms

Screen Shot 2022-07-15 at 5 14 58 PM

We have agreed not to hold up PRs due to undefined terms, but know that you are injecting a huge amount of undefined terms in this PR and in the future, we may resolve to exclude ANY credentials that do not define their terms fully from interoperability tests... this is because the graph processing value of credentials with poorly defined terms is drastically diminished, which is something I believe we are trying to support in the long term.

(We have similar issues in our schemas, but its a bad habit we all need to kick).

File: docs/openapi/components/schemas/common/FSMAShipment.yml
Comment (mprorock) [reply docs/openapi/components/schemas/common/FSMAShipment.yml:41]: Yes @rhofvendahl this should be corrected

File: docs/openapi/components/schemas/common/FSMAReceivingCTE.yml
Comment (mprorock) [reply docs/openapi/components/schemas/common/FSMAReceivingCTE.yml:131]: This is FDA language unfortunately

File: docs/openapi/components/schemas/common/FSMAReceivingCTE.yml
Comment (mprorock) [reply docs/openapi/components/schemas/common/FSMAReceivingCTE.yml:42]: +1

File: docs/openapi/components/schemas/common/FSMAProduct.yml
Comment (mprorock) [reply docs/openapi/components/schemas/common/FSMAProduct.yml:85]: A lot of this is because of PMI and FDA - we are not fans but need to support the regs as written

File: docs/openapi/components/schemas/common/FSMAProduct.yml
Comment (mprorock) [reply docs/openapi/components/schemas/common/FSMAProduct.yml:114]: That is straight from FDA

File: docs/openapi/components/schemas/common/FSMAGrowingCTE.yml
Comment (mprorock) [reply docs/openapi/components/schemas/common/FSMAGrowingCTE.yml:72]: Should use examp.org

File: docs/openapi/components/schemas/common/FSMACreatingCTE.yml
Comment (mprorock) [reply docs/openapi/components/schemas/common/FSMACreatingCTE.yml:66]: In this case FDA uses Unit code this way - we should say `FSMAUnitCode` or something like that @rhofvendahl

Comment (OR13) [reply docs/openapi/components/schemas/common/FSMACreatingCTE.yml:66]: How big is the coding system? We can define them all here possibly.
Or point to an existing ontology that defines them...

No hits in https://www.gs1.org/voc/

Have you checked https://qudt.org/ ?

**Comment by nissimsan** [comment]
@rhofvendahl the `FSMA` acronym isn't spelled out anywhere. I suggest you spell it out on the various schema `description`s. 
Great work!  :)

File: docs/openapi/components/schemas/common/FSMACreatingCTE.yml
Comment (rhofvendahl) [reply docs/openapi/components/schemas/common/FSMACreatingCTE.yml:66]: @OR13 the units are a little fuzzy, e.g. "crates", "pallets", "field bins".

@mprorock  correct me if I'm wrong but I think while standardized units for weight or volume might be useful, they're not currently part of the suggested reporting workflow (I'm guessing because users might not have easy access to that info &/or reviewers might not find that info particularly useful).

Looks like `https://schema.org/unitText` provides an alternative for units that don't have a standard unit code, which seems great for our purposes. Instead of referring to a QuantitativeValue I'll just provide `quantity` and `unit` properties where unit has id `unitText`.

File: docs/openapi/components/schemas/common/FSMAFirstReceiverData.yml
Comment (rhofvendahl) [reply docs/openapi/components/schemas/common/FSMAFirstReceiverData.yml:27]: My understanding is that hasBatchLotNumber is a string property attached to something like a product, whereas this traceabilityLot property is a ref to a bunch of lot data.

I do use hasBatchLotNumber as the id of the lotCode property within FSMATraceabilityLot

File: docs/openapi/components/schemas/common/FSMAGrowingCTE.yml
Comment (rhofvendahl) [reply docs/openapi/components/schemas/common/FSMAGrowingCTE.yml:28]: Yikes, thanks 😬

File: docs/openapi/components/schemas/common/FSMAGrowingCTE.yml
Comment (rhofvendahl) [reply docs/openapi/components/schemas/common/FSMAGrowingCTE.yml:72]: It is not a real email but I'll use produce@example.com to make that clear

File: docs/openapi/components/schemas/common/FSMAReceivingCTE.yml
Comment (rhofvendahl) [reply docs/openapi/components/schemas/common/FSMAReceivingCTE.yml:131]: Not very important but FWIW it's also the most clear language I can think of, or else I'd recommend shifting it 🤷‍♂️

`to` and `from` are properties on the FSMAShipment schema, which has a lot of data common to both `FSMAReceivingCTE` and `FSMAShippingCTE`. In both cases the `to` and `from` refer to the immediate previous destination and immediate subsequent destination of that leg of the journey.

The wording of the KDEs for both shipping & receiving is super clear about _immediate_ previous/subsequent locations, presumably so that the "source" for a product that's been shipped a lot is taken to mean the previous stop rather than the farm it was harvested on or something. So, whatever wording used has to imply this.

Given that `shipment` is used for KDEs shared between `shipping` and `receiving`, to/from seemed more clear than immediatePreviousSource/immediateNextDestination, or immediateSource/immediateDestination.

File: docs/openapi/components/schemas/common/FSMAShipment.yml
Comment (rhofvendahl) [reply docs/openapi/components/schemas/common/FSMAShipment.yml:41]: @OR13 Thanks for this info! I definitely let laziness get the best of me with id/type in this PR, I'll make that a priority going forward.

**Comment by rhofvendahl** [comment]
@OR13 thanks for bringing attention to the undefined terms issue. Regarding suggested fixes I'll submit changes to the minor stuff and leave bigger changes to subsequent PRs, as you suggest.

@nissimsan Thanks for the tip, I've added that info to certificate descriptions only to keep things succinct :-)

File: docs/openapi/components/schemas/common/FSMAGrowingCTE.yml
Comment (rhofvendahl) [reply docs/openapi/components/schemas/common/FSMAGrowingCTE.yml:70]: @OR13 This example text represents a reference to our current Entity schema. I'm not sure I'm following, what about that schema will result in undefined terms?

File: docs/openapi/components/schemas/common/FSMAReceivingCTE.yml
Comment (rhofvendahl) [reply docs/openapi/components/schemas/common/FSMAReceivingCTE.yml:74]: @OR13 This is another case where I am trying to reference our Entity schema. Is there a better way to do this?

**Review by mkhraisha — APPROVED** [review]

**Pull request merged**
